File: Projects/donut.java
public class Donut {
    public static void main(String[] args) {
        double A = 0, B = 0;  // Angles for rotation
        double i, j;
        
        while (true) {
            // Clear the screen
            System.out.print("\u001b[H");

            // Create 3D donut shape using ASCII characters
            for (i = 0; i < 6.28; i += 0.07) {  // Loop over latitude
                for (j = 0; j < 6.28; j += 0.02) {  // Loop over longitude
                    double sinA = Math.sin(A);
                    double cosA = Math.cos(A);
                    double sinB = Math.sin(B);
                    double cosB = Math.cos(B);
                    double sinI = Math.sin(i);
                    double cosI = Math.cos(i);
                    double sinJ = Math.sin(j);
                    double cosJ = Math.cos(j);

                    // 3D parametric equations for the donut shape
                    double x = cosA * (22 + 10 * cosI) * cosJ - sinA * sinJ;
                    double y = sinA * (22 + 10 * cosI) * cosJ + cosA * sinJ;
                    double z = 10 * sinI + 15;
                    double ooz = 1 / z;  // Perspective division

                    // Calculate screen position
                    int X = (int) (40 + 30 * x * ooz);
                    int Y = (int) (12 + 12 * y * ooz);

                    // Use brightness based on z value for better visualization
                    int L = (int) (8 * ((cosB * sinA * sinI - cosA * sinB * cosI) * cosJ - cosA * sinB * cosI - sinA * sinB * cosJ));
                    if (Y >= 0 && Y < 24 && X >= 0 && X < 80 && ooz > 0) {
                        // Map L value to ASCII brightness (32 is space, 46 is dot, 48 is zero, etc.)
                        char c = ".,-~:;=!*#$@"[Math.max(L, 0)];
                        System.out.print(c);
                    } else {
                        System.out.print(" ");
                    }
                }
                System.out.println();
            }

            // Increment rotation angles for the donut
            A += 0.04;
            B += 0.02;

            // Pause to simulate animation
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: Projects/donut.js
function donut() {
    let A = 0, B = 0;  // Angles for rotation
    const width = 80;
    const height = 24;
    const K2 = 5;  // Constant for zooming
    const K1 = width * K2 * 3 / (8 * (2 + 1));  // Constant for scaling
    
    setInterval(() => {
        // Create the 2D donut
        let output = Array(height).fill('').map(() => Array(width).fill(' '));
        
        for (let i = 0; i < 6.28; i += 0.07) {  // Loop over latitude
            for (let j = 0; j < 6.28; j += 0.02) {  // Loop over longitude
                let sinA = Math.sin(A);
                let cosA = Math.cos(A);
                let sinB = Math.sin(B);
                let cosB = Math.cos(B);
                let sinI = Math.sin(i);
                let cosI = Math.cos(i);
                let sinJ = Math.sin(j);
                let cosJ = Math.cos(j);

                // 3D parametric equations for the donut shape
                let x = cosA * (22 + 10 * cosI) * cosJ - sinA * sinJ;
                let y = sinA * (22 + 10 * cosI) * cosJ + cosA * sinJ;
                let z = 10 * sinI + 15;
                let ooz = 1 / z;  // Perspective division
                
                // Calculate screen position
                let X = Math.floor(width / 2 + K1 * x * ooz);
                let Y = Math.floor(height / 2 - K1 * y * ooz);
                
                // Calculate brightness based on z value for better visualization
                let L = Math.floor(8 * ((cosB * sinA * sinI - cosA * sinB * cosI) * cosJ - cosA * sinB * cosI - sinA * sinB * cosJ));
                if (Y >= 0 && Y < height && X >= 0 && X < width && ooz > 0) {
                    // Map L value to ASCII brightness (32 is space, 46 is dot, 48 is zero, etc.)
                    const chars = ".,-~:;=!*#$@";
                    const char = chars[Math.max(L, 0)];
                    output[Y][X] = char;
                }
            }
        }
        
        // Print the donut in the console
        console.clear();
        output.forEach(row => console.log(row.join('')));

        // Increment rotation angles for the donut
        A += 0.04;
        B += 0.02;

    }, 50);
}

// Start the donut animation
donut();
